userland/aero_ipc/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (C) 2021-2022 The Aero Project Developers.
+ *
+ * This file is part of The Aero Project.
+ *
+ * Aero is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Aero is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Aero. If not, see <https://www.gnu.org/licenses/>.
+ */
+#![feature(decl_macro)]
+
+use aero_syscall::{sys_ipc_recv, sys_ipc_send};
+use core::ops::DerefMut;
+use core::sync::atomic::{AtomicUsize, Ordering};
+use lazy_static::lazy_static;
+use serde::Deserialize;
+
+/// This is an ugly hack that exposes postcard and serde for the macro to use
+pub mod ipcmodules {
+    pub use postcard;
+    pub use serde;
+}
+
+/// A MessageHandler is a trait describing an IPC client
+pub trait MessageHandler: Send + Sync {
+    fn handle(&mut self, src: usize, msg: &[u8]) -> Option<Vec<u8>>;
+}
+
+/// A MessageTransport allows for high-level IPC exchanges over the IPC interfce.
+/// It also handles the allocation of request identifiers.
+pub trait MessageTransport {
+    fn alloc_id() -> usize;
+    fn free_id(id: usize);
+    fn exchange(meta: usize, mid: usize, data: &[u8]) -> Vec<u8>;
+}
+
+/// A SendRecieveTransport transfers messages by using the IPC system calls.
+pub struct SendRecieveTransport;
+static IDALLOC: AtomicUsize = AtomicUsize::new(0);
+impl MessageTransport for SendRecieveTransport {
+    fn alloc_id() -> usize {
+        return IDALLOC.fetch_add(1, Ordering::SeqCst);
+    }
+    fn free_id(_: usize) {}
+    fn exchange(meta: usize, mid: usize, msg: &[u8]) -> Vec<u8> {
+        // send the data
+        sys_ipc_send(meta, msg).expect("exchange failed: request failed!");
+        // now wait for a repsonse
+        loop {
+            // get a response
+            let rx = service_with_response_finding();
+            match rx {
+                // if we got a response,
+                Some(mut msg) => {
+                    // and the response has the correct message ID...
+                    let mut deser = postcard::Deserializer::from_bytes(&msg);
+                    let msgid = usize::deserialize(&mut deser)
+                        .expect("message ID not present in the message!");
+                    if msgid == (mid << 1) | 1 {
+                        // return the message contents!
+                        return msg.split_off(8);
```

Comment:
I would suggest you change that to `core::mem::size_of::<usize>()` or explicitly deserialize it as `u64` few lines above just to rule out a possible bug in the future where we try to run this code on a platform that isn't 64 bit :)

Reply:
yeah fair enough